### ISSUE TITLE
investigate the impact of Twisted 22.1

### ISSUE THREAD

**Issue by richvdh** [matrix-org/synapse#11848]
Announced at https://mail.python.org/archives/list/twisted@python.org/message/LJU54CY4S5GIOE46B7XCBVXONRV4BHBX/

> This is mainly a bugfix release.
>
> The main bug is:
> - twisted.web.client.RedirectAgent and twisted.web.client.BrowserLikeRedirectAgent now properly remove sensitive headers when redirecting to a different origin. (`#10294`)

This refers to https://twistedmatrix.com/trac/ticket/10294, which refers to https://twistedmatrix.com/trac/ticket/10295. In turn, that refers to https://github.com/twisted/twisted/pull/1683, which removes deprecated functions and notes GHSA-92x2-jw7w-xvvx. At the time of writing, that security advisory has not been published.

However, the [commit corresponding to the advisory](https://github.com/twisted/twisted/commit/af8fe78542a6f2bf2235ccee8158d9c88d31e8e2) is available, and it notes a list of [five sensitive headers](https://github.com/twisted/twisted/commit/af8fe78542a6f2bf2235ccee8158d9c88d31e8e2#diff-e4b69c8a7d3ea27ae42bd8b9ce6000b391e3901c61294ccf2d238ec102d458e9R2117-R2121).

**Speculation:** my (DMR) guess is that the deprecated functions and RedirectAgents were both forwarding sensitive headers. The former has been removed, and the latter has been fixed.

**Comment by DMRobertson** [comment]
twisted/twisted#1683's changelog:

> twisted.web.client.getPage, twisted.web.client.downladPage, and the associated implementation classes (HTTPPageGetter, HTTPPageDownloader, HTTPClientFactory, HTTPDownloader) have been removed because they do not segregate cookies by domain. They were deprecated in Twisted 16.7.0 in favor of twisted.web.client.Agent. GHSA-92x2-jw7w-xvvx.

Do we use any of these? `rg 'getPage|downloadPage|HTTPPAgeGetter|HTTPPageDownloader|HTTPClientFactory|HTTPDownloader'` yields no matches in sydent or sygnal. There's a match in synapse, but it's in an unrelated piece of javascript.

**Comment by DMRobertson** [comment]
As for the redirect agents:

> * twisted.web.client.RedirectAgent and twisted.web.client.BrowserLikeRedirectAgent now properly remove sensitive headers when redirecting to a different origin. (`#10294`)

It looks like we only use these for making `.well-known` requests:

```
2022-01-28 15:00:12 ✗ 130 INT  $ rg RedirectAgent synapse sydent sygnal
synapse/synapse/http/federation/well_known_resolver.py
24:from twisted.web.client import RedirectAgent
104:        self._well_known_agent = RedirectAgent(agent)

sydent/sydent/http/matrixfederationagent.py
31:from twisted.web.client import URI, Agent, HTTPConnectionPool, RedirectAgent
118:        self._well_known_agent = RedirectAgent(_well_known_agent)

sydent/stubs/twisted/web/client.pyi
116:class RedirectAgent:
```

The details have not been announced, but the corresponding commit is publicly [available](https://github.com/twisted/twisted/commit/af8fe78542a6f2bf2235ccee8158d9c88d31e8e2). 

In [Sydent](https://github.com/matrix-org/sydent/blob/143a22520b974ac63987de8dc048891071f299e4/sydent/http/matrixfederationagent.py#L343) we don't send any headers with this request. AFAICS Twisted doesn't add any of its own either.

In Synapse, we send a user agent header, but nothing else:

https://github.com/matrix-org/synapse/blob/bdfde6dca11a9468372b3c9b327ad3327cbdbe4a/synapse/http/federation/well_known_resolver.py#L245-L259

So I don't think we are sending any sensitive information in headers which would fall privy to the bug in twisted.

**Comment by DMRobertson** [comment]
No other removals in the [changelog](https://github.com/twisted/twisted/commit/af8fe78542a6f2bf2235ccee8158d9c88d31e8e2), and no other highlighted bugs.

TL;DR I think we're okay. But I would like to wait and see for the security advisory nonetheless.

**Comment by clokep** [comment]
> It looks like we only use these for making `.well-known` requests:

Note that treq also uses these internally: https://github.com/twisted/treq/blob/d89d553240f1c1e434bb137e5ce53877c54285e6/src/treq/client.py#L203-L207

So we would want to check any requests using that as well!

**Comment by DMRobertson** [comment]
`rg treq sygnal` yields no hits.

Only one hit in [Sydent](https://github.com/matrix-org/sydent/blob/48d2e7fb04539fa6174fba9cad123d6010b453ec/sydent/http/httpcommon.py#L153-L165) which doesn't look related.

Synapse, outside of tests:

```
\rg treq synapse/synapse -g '!tests'
synapse/synapse/python_dependencies.py
66:    "treq>=15.1",

synapse/synapse/rest/media/v1/preview_url_resource.py
140:            treq_args={"browser_like_redirects": True},

synapse/synapse/http/client.py
33:import treq
303:        treq_args: Optional[Dict[str, Any]] = None,
311:            treq_args: Extra keyword arguments to be given to treq.request.
323:        self._extra_treq_args = treq_args or {}
420:                request_deferred: defer.Deferred = treq.request(
426:                    # Avoid buffering the body in treq since we do not reuse
429:                    **self._extra_treq_args,
432:                # we use our own timeout mechanism rather than treq's as a workaround
860:            # stolen from https://github.com/twisted/treq/pull/49/files

synapse/synapse/http/matrixfederationclient.py
38:import treq
561:                            # that treq.request gives: just use the raw Agent.
616:                        d = treq.content(response)
```

**Comment by DMRobertson** [comment]
Chasing through the source, the first leads to `SimpleHttpClient.request`, which is the second hit. SimpleHttpClient is gonna get used all over the place.

The third isn't using `treq` to make a request, so I think it's fine.

**Comment by DMRobertson** [comment]
Let's go through each use of SimpleHttpClient. To start with, `get_proxied_blacklisted_http_client`. That's used in `FederationHandler` in a `get_json` call without any headers. It's also used in `HttpPusher` in two `post_json_get_json` calls, neither of which provide any headers.

`get_proxied_http_client` has four uses. Looking through these, I wasn't sure about 
- [this call](https://github.com/matrix-org/synapse/blob/a15a893df8428395df7cb95b729431575001c38a/synapse/handlers/oidc.py#L560-L574), because I don't know what headers authlib will add to `raw_headers`
- [this call](https://github.com/matrix-org/synapse/blob/a15a893df8428395df7cb95b729431575001c38a/synapse/handlers/oidc.py#L654-L657) sends `Authorization`, which _is_ one of the tokens Twisted considers sensitive.
I think these could use a sanity check from someone familiar with OIDC.

**Comment by richvdh** [comment]
There are two aspects to this task:

1. Check that Synapse is compatible with the forthcoming release.

    This is likely largely a no-op, since we have CI which checks nightly that our unit tests, and Sytest, run against twisted's trunk. Nevertheless it's good to double-check, and to that end I have updated a homeserver installation to have twisted 22.1.0rc1, and confirmed that it operates normally.

2. Since 22.1.0rc1 is a security fix, we need to evaluate to what extent it affects Synapse, so that we can in turn advise our users how urgent it is that they upgrade.

   @DMRobertson has begun that work above, with the conclusions that Synapse certainly does use the affected functionality, but, until the Twisted team publish their security advisory, it remains unclear how much that affects us. Accordingly, we are going to await release of that security advisory before progressing further.

**Comment by babolivier** [comment]
After syncing with @DMRobertson, we've concluded that Synapse isn't affected (in that we use the affected component but not in a way that can cause a security issue).
